lib/proxy/helpers.rb

```diff
@@ -80,7 +80,11 @@ def resolv(*args)
 
   # reverse lookup an IP address while verifying it via forward resolv
   def remote_fqdn(forward_verify = true)
-    ip = request.env['REMOTE_ADDR']
+    if request.env['HTTP_X_FORWARDED_FOR']
+      ip = request.env['HTTP_X_FORWARDED_FOR'].split(',')[0]
+    else
+      ip = request.env['REMOTE_ADDR']
+    end
```

Comment:
The major issue with blindly accepting `X-Forwarded-For` is that you're opening up a massive security issue since anyone can spoof these headers. That means you need proper middleware to secure the requests.

This weekend I was looking at Rack and I think [request.ip](https://rubydoc.info/github/rack/rack/Rack/Request/Helpers#ip-instance_method) should do the same. It appears to have an ip filter that you can set. I'd consider the default there insecure since it accepts RFC1918 IPs as well, which is common for a Smart Proxy deployment.

I'd say that an explicit list of allowed reverse proxies is more secure.

Also note that when using HTTPS client certificates (as you should) this is only used over HTTP. In a secure deployment, all authenticated modules should only be accessible over HTTPS.

I was actually thinking about proposing to drop this altogether.

Reply:
Yeah, this is a massive hole. We can't merge this as-is. I agree that we need to simplify SP's security.

What exactly do you need this for? Maybe we can help finding a better alternative.

Reply:
@lzap : We need to place the foreman smart proxy behind a reverse proxy. We would much rather check the forwarded certificate, like in foreman, but i don't know that much about your source code/ruby. This was a dirty try to get reverse proxy support and i appreciate, not the securest one. 
Would the verification of the forwarded certificate like in foreman be a secure alternative?

Reply:
>  Would the verification of the forwarded certificate like in foreman be a secure alternative?

Yes

Reply:
@ekohl : ok, and what have to be done to implement this?

Reply:
You can generate valid SSL client cert for the reverse proxy and add it to the "trusted hosts" configuration. All requests coming from this client will be authorized. If this is what you want.

Reply:
Currently thats our "workaround". But all clients, which can access or talk to the reverse proxy, are verified to the backend. Ok, i can add an require to some IP, but this isn't a very nice solution.
So, if you say the verification of an forwareded certificate is better, let me have a secound try. i will update the code.

Reply:
Then we could actually accept your patch but only with a new (required) setting - list of X509 common names that are allowed to resped x-forwarded-for. This way administrator must explicitly state which client cert is allowed to do this dangerous action. @ekohl what you think? Other way?